cpp/tests/interop/to_arrow_device_test.cpp

```diff
@@ -0,0 +1,735 @@
+    }
+  }
+};
+
+struct ToArrowDeviceTest : public BaseArrowFixture {};
+
+template <typename T>
+struct ToArrowDeviceTestDurationsTest : public BaseArrowFixture {};
+
+TYPED_TEST_SUITE(ToArrowDeviceTestDurationsTest, cudf::test::DurationTypes);
+
+TEST_F(ToArrowDeviceTest, EmptyTable)
+{
+  const auto [table, schema, arr] = get_nanoarrow_tables(0);
+
+  auto struct_meta          = cudf::column_metadata{"f"};
+  struct_meta.children_meta = {{"integral"}, {"string"}};
+
+  cudf::dictionary_column_view dview{table->view().column(2)};
+
+  auto got_arrow_schema =
+    cudf::to_arrow_schema(table->view(), {{"a"}, {"b"}, {"c"}, {"d"}, {"e"}, struct_meta});
+
+  compare_schemas(schema.get(), got_arrow_schema.get());
+  ArrowSchemaRelease(got_arrow_schema.get());
+
+  auto got_arrow_device = cudf::to_arrow_device(std::move(*table));
+  ASSERT_EQ(rmm::get_current_cuda_device().value(), got_arrow_device->device_id);
+  ASSERT_EQ(ARROW_DEVICE_CUDA, got_arrow_device->device_type);
+
+  compare_arrays(schema.get(), arr.get(), &got_arrow_device->array);
+  ArrowArrayRelease(&got_arrow_device->array);
+}
+
+TEST_F(ToArrowDeviceTest, DateTimeTable)
+{
+  auto data = {1, 2, 3, 4, 5, 6};
+  auto col =
+    cudf::test::fixed_width_column_wrapper<cudf::timestamp_ms, cudf::timestamp_ms::rep>(data);
+  std::vector<std::unique_ptr<cudf::column>> cols;
+  cols.emplace_back(col.release());
+  cudf::table input(std::move(cols));
+
+  auto got_arrow_schema = cudf::to_arrow_schema(input.view(), {{"a"}});
+  nanoarrow::UniqueSchema expected_schema;
+  ArrowSchemaInit(expected_schema.get());
+  ArrowSchemaSetTypeStruct(expected_schema.get(), 1);
+  ArrowSchemaInit(expected_schema->children[0]);
+  ArrowSchemaSetTypeDateTime(
+    expected_schema->children[0], NANOARROW_TYPE_TIMESTAMP, NANOARROW_TIME_UNIT_MILLI, nullptr);
+  ArrowSchemaSetName(expected_schema->children[0], "a");
+  expected_schema->children[0]->flags = 0;
+
+  compare_schemas(expected_schema.get(), got_arrow_schema.get());
+  ArrowSchemaRelease(got_arrow_schema.get());
+
+  auto data_ptr        = input.get_column(0).view().data<int64_t>();
+  auto got_arrow_array = cudf::to_arrow_device(std::move(input));
+  ASSERT_EQ(rmm::get_current_cuda_device().value(), got_arrow_array->device_id);
+  ASSERT_EQ(ARROW_DEVICE_CUDA, got_arrow_array->device_type);
+
+  ASSERT_EQ(data.size(), got_arrow_array->array.length);
+  ASSERT_EQ(0, got_arrow_array->array.null_count);
+  ASSERT_EQ(0, got_arrow_array->array.offset);
+  ASSERT_EQ(1, got_arrow_array->array.n_children);
+  ASSERT_EQ(nullptr, got_arrow_array->array.buffers[0]);
+
+  ASSERT_EQ(data.size(), got_arrow_array->array.children[0]->length);
+  ASSERT_EQ(0, got_arrow_array->array.children[0]->null_count);
+  ASSERT_EQ(0, got_arrow_array->array.children[0]->offset);
+  ASSERT_EQ(nullptr, got_arrow_array->array.children[0]->buffers[0]);
+  ASSERT_EQ(data_ptr, got_arrow_array->array.children[0]->buffers[1]);
+
+  ArrowArrayRelease(&got_arrow_array->array);
```

Comment:
```suggestion
  EXPECT_EQ(rmm::get_current_cuda_device().value(), got_arrow_array->device_id);
  EXPECT_EQ(ARROW_DEVICE_CUDA, got_arrow_array->device_type);

  EXPECT_EQ(data.size(), got_arrow_array->array.length);
  EXPECT_EQ(0, got_arrow_array->array.null_count);
  EXPECT_EQ(0, got_arrow_array->array.offset);
  EXPECT_EQ(1, got_arrow_array->array.n_children);
  EXPECT_EQ(nullptr, got_arrow_array->array.buffers[0]);

  EXPECT_EQ(data.size(), got_arrow_array->array.children[0]->length);
  EXPECT_EQ(0, got_arrow_array->array.children[0]->null_count);
  EXPECT_EQ(0, got_arrow_array->array.children[0]->offset);
  EXPECT_EQ(nullptr, got_arrow_array->array.children[0]->buffers[0]);
  EXPECT_EQ(data_ptr, got_arrow_array->array.children[0]->buffers[1]);

  ArrowArrayRelease(&got_arrow_array->array);
```
I'm not sure if these warrant an assert. I'm a bit worried that triggering an assert would cause a memory leak since the test would be stopped before the `ArrowArrayRelease` gets called.

Reply:
I wasn't as particularly worried about the memory leak since it was a test, but it's a good point.

Reply:
This pattern appears in other gtests as well.

Reply:
I've updated the other tests to use `EXPECT`